src/kernels/cosine.jl

```diff
@@ -0,0 +1,14 @@
+"""
+    CosineKernel
+
+The cosine kernel is a stationary kernel for a sinusoidal given by
+```
+    κ(x,y) = cos( 2π * (x-y) )
```

Comment:
Since `CosineKernel` uses the `Cityblock` distance, I guess it should rather be
```suggestion
κ(x, y) = cos(2π * ‖x-y‖₁)
```

Reply:
Actually is it really the city block metric? I would expect to see the euclidean one. Actually is there a reference for this kernel? That would be nice to also have it in the documentation

Reply:
I haven't been able to find a proper reference for the Cosine kernel. The best I could find was GPML's [official manual](http://www.gaussianprocess.org/gpml/code/matlab/doc/manual.pdf) pg44.

`Cityblock` metric is something I inferred from [here](https://github.com/alshedivat/gpml/blob/7af284c421804d9cd00a77294c1bbb092546b40a/cov/covCos.m#L34).

GPML's dosctring says `(x-y)`. Please note that cosine is an even function.

Reply:
Hmmm, actually, does the `Cosine` kernel actually make sense in > 1D? i.e. if you use the `CityBlock` metric, is it positive definite?

Reply:
GMPL [only defines the cosine kernel for one-dimensional data](https://github.com/alshedivat/gpml/blob/7af284c421804d9cd00a77294c1bbb092546b40a/cov/covCos.m#L22-L23) (also stated in the manual), and hence clearly `x-y` and `|x-y|` yield the same values due to the symmetry of cosine. I agree with @theogf that probably the most natural extension to multiple dimensions would be the Euclidean distance but unfortunately I don't have any reference at hand.

Reply:
GPyTorch uses the Euclidean distance: https://gpytorch.readthedocs.io/en/latest/kernels.html#cosinekernel

Reply:
Also note that GPyTorch uses `pi` instead of `2*pi` inside the cos term.

Reply:
Actually, according to the GPML docs GPML should use `pi` as well (but it seems in the implementation they use `2*pi`).

Reply:
Just reiterating my point above, the following suggests it doesn't make sense in higher dimensions:
```julia
using Stheno
using Stheno: Euclidean
import Stheno: pw

pw(k::Cosine, x::ColVecs) = cos.(pi .* pw(Euclidean(), x.X) ./ k.p)

x = ColVecs(randn(3, 10))
eigvals(pw(Cosine(1.0), x))
10-element Array{Float64,1}:
 -2.6807114471460594
 -1.3962293964372774
 -0.963215712851525
 -0.40797823094040286
  0.9649993076661302
  1.6352424163325816
  1.8341656854612294
  2.543664324231879
  3.937344284513905
  4.53271876916954
```
Obviously you'll get different numbers than me because seeds, but looks like it doesn't make sense.

Reply:
From what I looked very quickly, the cosine kernel support is only on [0,1] in 1D, which could be expanded to the unit sphere for higher dimensions

Reply:
Oh, weird. Does it not extend to the entire real line by virtue of the periodicity of `cos`?

Reply:
So, do you suggest we extend it to multi-dim sphere?

Reply:
I don't know that we have the right abstractions to handle this at the minute. I would suggest sticking with the real line so that we can continue to push forwards.